File: core/src/io/github/jerukan/game/ui/gamescreen/menus/UnitBuildMenu.java
package io.github.jerukan.game.ui.gamescreen.menus;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.utils.Align;
import io.github.jerukan.game.GameState;
import io.github.jerukan.game.WorldRenderer;
import io.github.jerukan.game.board.BoardManager;
import io.github.jerukan.game.gameunits.UnitRegistry;
import io.github.jerukan.game.gameunits.unitdata.BaseUnit;
import io.github.jerukan.game.ui.ButtonGroup;
import io.github.jerukan.game.ui.gamescreen.buttons.UnitBuildButton;
import io.github.jerukan.util.NamedFlag;
import io.github.jerukan.util.Constants;
import io.github.jerukan.util.Position;

import java.util.ArrayList;

/** Menu that appears when a player is selecting a unit to build */

public class UnitBuildMenu extends ButtonGroup {

    private Position currentPos;
    private ArrayList<UnitBuildButton> unitButtons;

    public UnitBuildMenu(NamedFlag[] flags) {
        super(flags);
        currentPos = new Position();
        table.align(Align.bottomLeft);
        unitButtons = new ArrayList<>();
        for(BaseUnit unit : UnitRegistry.unitList) {
            unitButtons.add(new UnitBuildButton(unit, flagFromArray("build")));
        }
        resetTable();
    }

    public void resetTable() {
        table.clear();
        for(UnitBuildButton button : unitButtons) {
            if(button.getBaseUnit()._canBuild(BoardManager.getSelectedPosition(), GameState.instance.getCurrentPlayer())) {
                table.add(button).pad(5).row();
            }
        }
    }

    public UnitBuildButton getHoveredButton() {
        for(UnitBuildButton button : unitButtons) {
            if(button.isHovered()) {
                return button;
            }
        }
        return null;
    }

    @Override
    public void updateVisibility() {
        if(!currentPos.equals(BoardManager.getSelectedPosition())) {
            flagFromArray("build").setState(false);
            currentPos.setPos(BoardManager.getSelectedPosition());
        }
        if(flagFromArray("build").getState()) {
            if(BoardManager.getSelectedPosition().isValid()) {
                if (GameState.instance.unitState.positionAvailable(BoardManager.getSelectedPosition())) {
                    WorldRenderer.boardCam.updateOffsets();
                    Sprite s = GameState.instance.boardState.tileFromPosition(BoardManager.getSelectedPosition()).getSprite();
                    table.setPosition(s.getX() + s.getWidth() + Constants.TILE_MENU_OFFSET - WorldRenderer.boardCam.camOffsetX, s.getY() - WorldRenderer.boardCam.camOffsetY);
                    table.setVisible(true);
                }
                else {
                    flagFromArray("build").setState(false);
                }
            }
        }
        else {
            resetTable();
            table.setVisible(false);
        }
    }
}

File: core/src/io/github/jerukan/game/ui/gamescreen/menus/UnitBuildMenu.kt
package io.github.jerukan.game.ui.gamescreen.menus

import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.utils.Align
import io.github.jerukan.game.GameState
import io.github.jerukan.game.WorldRenderer
import io.github.jerukan.game.gameunits.UnitRegistry
import io.github.jerukan.game.gameunits.unitdata.BaseUnit
import io.github.jerukan.game.ui.ButtonGroup
import io.github.jerukan.game.ui.gamescreen.buttons.UnitBuildButton
import io.github.jerukan.util.NamedFlag
import io.github.jerukan.util.Constants
import io.github.jerukan.util.Position

import java.util.ArrayList

/** Menu that appears when a player is selecting a unit to build  */

class UnitBuildMenu(flags: Array<NamedFlag>) : ButtonGroup(flags) {

    private val currentPos: Position
    private val unitButtons: ArrayList<UnitBuildButton>

    val hoveredButton: UnitBuildButton?
        get() {
            for (button in unitButtons) {
                if (button.isHovered) {
                    return button
                }
            }
            return null
        }

    init {
        currentPos = Position()
        table.align(Align.bottomLeft)
        unitButtons = ArrayList()
        for (unit in UnitRegistry.unitList) {
            unitButtons.add(UnitBuildButton(unit, flagFromArray("build")))
        }
        resetTable()
    }

    fun resetTable() {
        table.clear()
        for (button in unitButtons) {
            if (button.baseUnit._canBuild(GameState.instance.boardManager.selectedPosition, GameState.instance.currentPlayer)) {
                table.add(button).pad(5f).row()
            }
        }
    }

    override fun updateVisibility() {
        if (!currentPos.equals(GameState.instance.boardManager.selectedPosition)) {
            flagFromArray("build").state = false
            currentPos.setPos(GameState.instance.boardManager.selectedPosition)
        }
        if (flagFromArray("build").state) {
            if (GameState.instance.boardManager.selectedPosition.isValid) {
                if (GameState.instance.unitManager.positionAvailable(GameState.instance.boardManager.selectedPosition)) {
                    WorldRenderer.boardCam.updateOffsets()
                    val s = GameState.instance.boardManager.tileFromPosition(GameState.instance.boardManager.selectedPosition).sprite
                    table.setPosition(s.x + s.width + Constants.TILE_MENU_OFFSET - WorldRenderer.boardCam.camOffsetX, s.y - WorldRenderer.boardCam.camOffsetY)
                    table.isVisible = true
                } else {
                    flagFromArray("build").state = false
                }
            }
        } else {
            resetTable()
            table.isVisible = false
        }
    }
}
